### PR TITLE
[AIRFLOW-7044] Host key can be specified via SSH connection extras.

### PR DESCRIPTION
**Pull request by triptec**
Original PR: https://github.com/apache/airflow/pull/8536

This PR adds functionality that allows a RSA public key part to be specified in the extra field of the SSH Connection. This will improve security by making it easier to specify a host key to verify connections against. Previously, you had to edit the ~.ssh/known_hosts file directly which I have found cumbersome to maintain when running Airflow on Kubernetes.

### CODE REVIEW

**Comment by turbaszek** [comment]
@triptec can you please add meaningful title and description to your PR?

**Comment by triptec** [comment]
@turbaszek ah, I thought I made it against my own master, sorry

**Comment by triptec** [comment]
@turbaszek any thoughts about the PR?

File: tests/providers/ssh/hooks/test_ssh.py [hunk 1]
```diff
@@ -51,8 +51,18 @@ def generate_key_string(pkey: paramiko.PKey, passphrase: Optional[str] = None):
...
+    pkey.write_private_key(key_fh)
+    key_fh.seek(0)
+    key_obj = paramiko.RSAKey(file_obj=key_fh)
+    return key_obj
+
+
 TEST_PKEY = paramiko.RSAKey.generate(4096)
 TEST_PRIVATE_KEY = generate_key_string(pkey=TEST_PKEY)
+TEST_HOST_PKEY = generate_host_key(pkey=TEST_PKEY)
+TEST_HOST_KEY = TEST_HOST_PKEY.get_base64()
```
Comment (ashb) [inline tests/providers/ssh/hooks/test_ssh.py]: Why not just have `generate_host_key` return the base64 encoded version?

File: airflow/providers/ssh/hooks/ssh.py [hunk 2]
```diff
@@ -149,7 +151,9 @@ def __init__(
                     and str(extra_options["look_for_keys"]).lower() == 'false'
                 ):
                     self.look_for_keys = False
-
+                if "host_key" in extra_options and self.no_host_key_check is False:
+                    encoded_host_key = decodebytes(extra_options["host_key"].encode('utf-8'))
+                    self.host_key = paramiko.RSAKey(data=encoded_host_key)
```
Comment (ashb) [inline airflow/providers/ssh/hooks/ssh.py]:
```suggestion
                    decoded_host_key = decodebytes(extra_options["host_key"].encode('utf-8'))
                    self.host_key = paramiko.RSAKey(data=decoded_host_key)
```

File: airflow/providers/sftp/hooks/sftp.py [hunk 3]
```diff
@@ -109,12 +111,22 @@ def __init__(self, ftp_conn_id: str = 'sftp_default', *args, **kwargs) -> None:
                     )
                     self.key_file = extra_options.get('private_key')
 
+                if "host_key" in extra_options and self.no_host_key_check is False:
+                    encoded_host_key = decodebytes(extra_options["host_key"].encode('utf-8'))
+                    self.host_key = paramiko.RSAKey(data=encoded_host_key)
+
```
Comment (ashb) [inline airflow/providers/sftp/hooks/sftp.py]: This isn't needed here -- it's handled by SSHHook already.

File: docs/apache-airflow-providers-ssh/connections/ssh.rst [hunk 4]
```diff
@@ -47,9 +47,10 @@ Extra (optional)
...
     * ``allow_host_key_change`` - Set to ``true`` if you want to allow connecting to hosts that has host key changed or when you get 'REMOTE HOST IDENTIFICATION HAS CHANGED' error.  This wont protect against Man-In-The-Middle attacks. Other possible solution is to remove the host entry from ``~/.ssh/known_hosts`` file. Default is ``false``.
     * ``look_for_keys`` - Set to ``false`` if you want to disable searching for discoverable private key files in ``~/.ssh/``
+    * ``host_key`` - The base64 encoded ssh-rsa public key of the host. Specifying this, along with ``no_host_key_check=False`` allows you to only make the connection if the public key of the endpoint matches this value.
```
Comment (ashb) [inline docs/apache-airflow-providers-ssh/connections/ssh.rst]:
```suggestion
    * ``host_key`` - The base64 encoded ssh-rsa public key of the host, as you would find in the ``known_hosts`` file. Specifying this, along with ``no_host_key_check=False`` allows you to only make the connection if the public key of the endpoint matches this value.
```

**Review by ashb — CHANGES_REQUESTED** [review]

File: airflow/providers/sftp/hooks/sftp.py (hunk 3 above)
Comment (triptec) [reply airflow/providers/sftp/hooks/sftp.py]: Yeah, you're right, I missed that part.

**Comment by triptec** [comment]
@ashb I've made the changes and added a test. The ssh hook sets no_host_key_check to True by default but that is overridden to False by default in the sftp hook, but as no_host_key_check is set to False in sftp's __init__ after ssh's __init__ it means you have to set no_host_key_check: false in extras even if that's default in sftp.

**Comment by aaronfowles-emis** [comment]
Thanks for acknowledging the original PR @triptec. It would have been nice to get my commits in as I remember those tests taking quite a bit of my time to write. Main thing is glad to see you're progressing this though.

**Comment by triptec** [comment]
@aaronfowles-emis Yeah, it was the least I could do, I tried to work of your branch to keep the commits intact but got a lot more than just the changes in your PR so I ended up having to just make a patch of the diff between your master and your PR and apply that to the current master, only way I could solve it though I bet there are other ways. Do you have any suggestion of what you would like me to do?

**Comment by triptec** [comment]
@aaronfowles-emis, @aaronfowles how about I change the attribution of https://github.com/apache/airflow/pull/12944/commits/eb1516ee70d2a145c6e24de14c3d0d07437d16c3 to "aaron.fowles <aaron.fowles@ovoenergy.com>"? Would that be ok?

**Comment by aaronfowles-emis** [comment]
> @aaronfowles-emis, @aaronfowles how about I change the attribution of [eb1516e](https://github.com/apache/airflow/commit/eb1516ee70d2a145c6e24de14c3d0d07437d16c3) to "aaron.fowles [aaron.fowles@ovoenergy.com](mailto:aaron.fowles@ovoenergy.com)"? Would that be ok?

That would be awesome - thanks :). If you could change it to `aaronfowles@gmail.com` that would be even better. Many thanks.

**Comment by triptec** [comment]
@aaronfowles-emis Done.

**Review by ashb — APPROVED** [review]

**Comment by triptec** [comment]
@ashb Thanks!

**Comment by triptec** [comment]
@ashb should I continue to rebase or can I leave it as it is?

**Comment by potiuk** [comment]
Can you please rebase? We just fixed failing master

**Comment by triptec** [comment]
@potiuk Yeah I’ve rebased.

**Pull request merged**